Extend the Extraction Plugin to synthesise OCaml external and callback definitions for interfacing C/C++ (#18270)
coq/coq: doc/sphinx/addendum/extraction.rst

        Axiom Y : Set -> Set -> Set.
        Extract Constant Y "'a" "'b" => " 'a * 'b ".

   .. note::
      The extraction recognizes whether the realized axiom
      should become a ML type constant or a ML object declaration. For example:

      .. coqtop:: in

         Axiom X:Set.
         Axiom x:X.
         Extract Constant X => "int".
         Extract Constant x => "0".

   .. caution:: It is the responsibility of the user to ensure that the ML
      terms given to realize the axioms do have the expected types. In
      fact, the strings containing realizing code are just copied to the
      extracted files.

   .. exn:: The term @qualid is already defined as foreign custom constant.

      The :n:`@qualid` was previously used in a
      :cmd:`Extract Foreign Constant` command. Using :cmd:`Extract Inlined Constant`
      for :n:`@qualid` would override this command.

> 

Reviewer: Given the message, it  seems specific to `Extract Inlined Constant`, so it should be moved below.

Author: Indeed, that does make sense.

Author: Fixed.

      by a :cmd:`Extraction Inline`. Hence a :cmd:`Reset Extraction Inline`
      will have an effect on the realized and inlined axiom.

   .. caution:: It is the responsibility of the user to ensure that the ML
      terms given to realize the axioms do have the expected types. In
      fact, the strings containing realizing code are just copied to the
      extracted files. The extraction recognizes whether the realized axiom
      should become a ML type constant or a ML object declaration. For example:

   .. exn:: The term @qualid is already defined as foreign custom constant.

      The :n:`@qualid` was previously used in a
      :cmd:`Extract Foreign Constant` command. Using :cmd:`Extract Inlined Constant`
      for :n:`@qualid` would override this command.

> 

Reviewer: This breaks the documentation. Note that you can check the produced documentation by clicking on the link next to the job `doc:refman: refman artifact` in the list of checks at the bottom of the pull request.

First, the "caution" block is meant to apply to both commands `Extract Constant` and `Extract Inlined Constant`, so there is no reason to indent it here. Another possibility is to move the block under `Extract Constant`, since it is the primary command.

Second, you have put the "error" block right in the middle of the "caution" block (notice the "For example" part); you should move it earlier, e.g., before the "note" block.

Third (it is not your fault, but you might just as well fix it along the way), the sentence "The extraction recognizes..." should be moved outside the "caution" block, next to the Coq script below. It could also be made into a "note" block (including the Coq script) and moved under `Extract Constant`.

Author: Thank you @silene. I will fix that. Should I also squeeze the changes with my previous commit and/or rebase on the current state of coq/master?

Author: Okay, now the layout looks good and my branch is in sync with coq/master. I'll wait for the checks to complete. Is the full CI done automatically?

Author: Made some small changes on the FFI section. I had some plain @string and @qualid strings. They should also be references. I'll check the refman when it is finished. I cannot check offline since the matching sphinx version is not supported on my system.

   Same as the :cmd:`Extract Constant`, except that the given ML terms
   will not be declared via ``let`` but in the form
   ``external @qualid : ML type = "@string"``.
   Also, no ``val`` declaration will be generated in the respective
   mli file.

.. caution::

   * The external function name provided by the user as :n:`@string`
     is currently **not** checked at all by extraction, even for syntax errors.

   * It is the responsibility of the user to ensure that the C
     functions given to realize the axioms do have the expected or compatible
     types. In fact, the strings containing realizing code are just copied to the
     extracted files.

   * A previously used :cmd:`Extract Constant` for :n:`@qualid` will be overridden
     by the command :cmd:`Extract External Constant`, although internally, the
     specified :n:`@qualid` will still be part of the (inline) constant extraction
     table. This is by design, since, when defining a function to be external, a
     ``val`` declaration and ``let`` definition would collide with the external declaration
     during compilation. Also, an external function cannot be inlined. Thus, the

>      command :cmd:`Extract Inlined Constant` has to be overridden, too.

Reviewer: Does the user actually care?

In fact, I am not even sure what the point of this silent behavior is. Would it not be better to emit an error if the extraction scheme for a given constant suddenly changes?

Author: Here, I did not want to be too rude to the user and rather warn in the documentation. But if it is okay to be hard and handle that as error, it is okay for me. I adopted the code accordingly.

.. note::
   Since the generated external expression contains the mapped OCaml type, the
   C/C++ compiler is able to determine whether the ML function and respective
   C/C++ function match in the arity and whether the types are compatible. Thus,

>    the type safety is not undermined when using :cmd:`Extract External Constant`.

Reviewer: This whole paragraph is completely wrong; please remove it. Indeed, there is absolutely no check whatsoever, not even on the arity of the function. As long as there is a symbol corresponding to the string, the compilation/link will succeed it.

Author: Yes, it is. I tested that for passing not enough and too much parameters to an external function. The linker indeed doesn't care.

.. exn:: Extract External Constant is supported only for OCaml extraction

   Calling foreign functions is gravely different between OCaml, Haskell and
   Scheme. Thus, extracting externals does not make sense when extracting into

>    anything other than OCaml.

Reviewer: The fact that FFI is different between OCaml and Haskell and Scheme does not mean than it does not make sense to extract externals to Haskell or Scheme. In fact, the difference between these systems is almost negligible. In OCaml, you do `external name : type = "string"`, while in Haskell you would do `foreign import capi "string" name :: type`, and in Scheme you would do `define name (foreign-procedure "string" type)`.

By the way, this makes the whole "external" naming ill-suited for the feature. It would be better to have a more agnostic naming, e.g., `Extract Foreign Constant`.

Author: This is also correct. Again, sorry for that one. I read the documentation of Haskell and Scheme concerning FFIs. This should be possible with quite few changes to my code. For Haskell, I need an additional line at the begin of the file. Should I extend the support do Haskell and Scheme or restrict my changes to OCaml for now. Also I saw, that Haskell has `foreign export` which is comparable to the OCaml callbacks. But I don't see a well-fitting common term for those concepts - except from Extract Closure. But the term closure is quite overloaded.
For Scheme, I have to dig through the documentation concerning something comparable to OCaml callback registration.

Also, I like the proposal `Extract Foreign Constant` and adopted that.

   Extract Inductive nat => int [ "0" "succ" ] "(fun fO fS n -> if n=0 then fO () else fS (n-1))".

Generating FFI Code
~~~~~~~~~~~~~~~~~~~

The plugin provides mechanisms to generate (currently only) OCaml code to
interface the generated OCaml code with C programs. In order to link compiled
OCaml code with C code, the linker needs to know

 * which C functions will be called by the ML code (external)
 * which ML functions shall be accessible by the C code (callbacks)

.. cmd:: Extract Foreign Constant @qualid => @string

>    Same as the :cmd:`Extract Constant`, except that the given ML terms

Reviewer 2: ```suggestion
   Like :cmd:`Extract Constant`, except that the referenced ML terms
```

"Same, ... except" is a bit contradictory.
The command has the qualid of the term, not the term itself.

   will be declared in the form

   ``external @qualid : ML type = "@string"``.

.. caution::

>    * The external function name provided by the user as :n:`@string`
>      is currently **not** checked at all by extraction, even for syntax errors.
> 
>    * It is the responsibility of the user to ensure that the C
>      functions given to realize the axioms do have the expected or compatible
>      types. In fact, the strings containing realizing code are just copied to the
>      extracted files.

Reviewer 2: ```suggestion
   * The external function name :n:`@string` is not checked in any way.

   * The user must ensure that the C
     functions given to realize the axioms have the expected or compatible
     types. In fact, the strings containing realizing code are just copied to the
     extracted files.
```

.. exn:: Extract Foreign Constant is supported only for OCaml extraction

>    We currently only support calling foreign functions for OCaml.

Reviewer 2: ```suggestion
   Foreign function calls are only supported for OCaml.
```

.. exn:: The term @qualid is already defined as custom constant.

>    This error message means that the :n:`@qualid` was previously used in the command
>    :cmd:`Extract Constant` or :cmd:`Extract Inlined Constant`. Using :cmd:`Extract Foreign Constant`
>    for :n:`@qualid` would override those commands.

Reviewer 2: ```suggestion
   The :n:`@qualid` was previously used in a
   :cmd:`Extract Constant` or :cmd:`Extract Inlined Constant` command. Using :cmd:`Extract Foreign Constant`
   for :n:`@qualid` would override those commands.
```

How is the last sentence relevant to the error description?

.. cmd:: Extract Callback {? @string } @qualid

   This command makes sure that after extracting the :term:`constants <constant>`
   specified by :n:`@qualid`, a constant ML function will be generated that
   registers :n:`@qualid` as callback, callable by :n:`@string`.
   This is done by declaring a function
   ``let _ = Callback.register "@string" @qualid``
   This expression signals OCaml that the given ML function :n:`@qualid` shall be
   accessible via the alias :n:`@string`, when calling from C/C++.
   If no alias is specified, it is set to the string representation of :n:`@qualid`.

.. caution::

>    * The optional alias provided by the user as :n:`@string`
>      is currently **not** checked at all by extraction, even for syntax errors.

Reviewer 2: ```suggestion
   * The optional alias :n:`@string` is currently **not** checked in any way.
```

>    * It is the responsibility of the user to ensure that the callback aliases are

Reviewer 2: ```suggestion
   * The user must ensure that the callback aliases are
```

     unique, i.e. when multiple modules expose a callback, the user should make
     sure that no two :n:`@qualid` share the same alias.

.. note::
   Using Extract Callback has no impact on the rest of the synthesised code since

>    it is an additional declaration. Thus, there is also no impact on correctness

Reviewer 2: ```suggestion
   it is an additional declaration. Thus, there is no impact on the correctness
```

   and type safety of the generated code.

.. exn:: Extract Callback is supported only for OCaml extraction

>    The concept of callback registration via calling ``Callback.register`` is specific
>    for OCaml. Thus, the command is only usable when extracting OCaml code.

Reviewer 2: ```suggestion
   The callback registration mechanism ``Callback.register`` is specific
   to OCaml. Thus, the command is only usable when extracting OCaml code.
```

.. cmd:: Print Extraction Foreign

>    Prints the current state of the set recording the custom foreigns
>    declared by the command :cmd:`Extract Foreign Constant` together with
>    the defined foreign ML function name.

Reviewer 2: ```suggestion
   Prints the current set of custom foreign functions
   declared by the command :cmd:`Extract Foreign Constant` together with its
   associated foreign ML function name.
```

Is it printing the functions or just the names of the functions?

.. cmd:: Reset Extraction Foreign

>    Resets the current state of the set recording the custom externals
>    declared by the command :cmd:`Extract Foreign Constant`.

Reviewer 2: ```suggestion
   Resets the set of custom externals
   declared by the command :cmd:`Extract Foreign Constant`.
```

or "Clears ..."

Author: Resets is okay here, I think. This wording is consistent with the description of previously existing Reset commands

.. cmd:: Print Extraction Callback

>    Prints the current state of the map recording the callbacks
>    declared by the command :cmd:`Extract Callback`
>    by printing the respective :token:`qualid` and callback alias
>    :token:`string` or "no custom alias" if the optional :token:`string`
>    was not set for that :token:`qualid`.

Reviewer 2: ```suggestion
   Prints the map of callbacks
   declared by the command :cmd:`Extract Callback`,
   showing the :token:`qualid` and callback alias
   :token:`string` (if specified) for each callback.
```

.. cmd:: Reset Extraction Callback

>    Resets the current state of the map recording the callbacks

Reviewer 2: ```suggestion
   Resets the the map recording the callbacks
```

   Extract Inductive nat => int [ "0" "succ" ] "(fun fO fS n -> if n=0 then fO () else fS (n-1))".

Generating FFI Code
~~~~~~~~~~~~~~~~~~~

The plugin provides mechanisms to generate only OCaml code to
interface the generated OCaml code with C programs. In order to link compiled
OCaml code with C code, the linker needs to know

   * which C functions will be called by the ML code (external)
   * which ML functions shall be accessible by the C code (callbacks)

.. cmd:: Extract Foreign Constant @qualid => @string

   Like :cmd:`Extract Constant`, except that the referenced ML terms
   will be declared in the form

   ``external @qualid : ML type = "@string"``.

> .. caution::

Reviewer 2: ```suggestion
   .. caution::
```

`.. caution`, `.. exn` and `.. warn` blocks should be indented 3 more spaces to match what we've done elsewhere in the doc.  Looks like there are 7 blocks to adjust.

Author: Okay, I will adjust that.

   accessible via the alias :n:`@string`, when calling from C/C++.
   If no alias is specified, it is set to the string representation of :n:`@qualid`.

.. caution::
   * The optional alias :n:`@string` is currently **not** checked in any way.

   * The user must ensure that the callback aliases are
     unique, i.e. when multiple modules expose a callback, the user should make
     sure that no two :n:`@qualid` share the same alias.

.. note::
   Using Extract Callback has no impact on the rest of the synthesised code since
   it is an additional declaration. Thus, there is no impact on the correctness
   and type safety of the generated code.

.. exn:: Extract Callback is supported only for OCaml extraction.

   The callback registration mechanism ``Callback.register`` is specific
   to OCaml. Thus, the command is only usable when extracting OCaml code.

.. cmd:: Print Extraction Foreign

   Prints the current set of custom foreign functions
   declared by the command :cmd:`Extract Foreign Constant` together with its
   associated foreign ML function name.

.. .. cmd:: Reset Extraction Foreign

..   Resets the set of custom externals

> ..   declared by the command :cmd:`Extract Foreign Constant`.

Reviewer 2: These are comments, not visible in the doc.  Is that intended and needed?

Author: Yes, @jfehrle, his is intended.
See my [comment](https://github.com/coq/coq/pull/18270/#issuecomment-1810268536). It is not clear yet, how to handle the resetting of (inlined/foreign) customs, i.e. removing a (global) substitution for a qualid. This is not a problem of this PR but rather a general one which should be addressed separately. Thus I deactivated the functionality and documentation for `Reset Extraction Foreign` for now.
